svm/src/account_loader.rs

```diff
@@ -477,13 +493,216 @@ struct LoadedTransactionAccounts {
+                    {
+                        // ...and the programdata account exists (if it doesn't, it is *not* a load failure)...
+                        if let Some(programdata_account) =
+                            account_loader.load_account(&programdata_address)
+                        {
+                            // ...count programdata toward this transaction's total size.
+                            loaded_transaction_accounts.increase_calculated_data_size(
+                                TRANSACTION_ACCOUNT_BASE_SIZE
+                                    .saturating_add(programdata_account.data().len()),
+                                loaded_accounts_bytes_limit,
+                                error_metrics,
+                            )?;
+                            additional_loaded_accounts.insert(programdata_address);
+                        }
+                    }
+                }
+            }
+
+            loaded_transaction_accounts.accounts.push((*key, account));
+
+            Ok(())
+        };
+
+    // Since the fee payer is always the first account, collect it first.
+    // We can use it directly because it was already loaded during validation.
+    collect_loaded_account(
+        account_loader,
+        message.fee_payer(),
+        loaded_fee_payer_account,
+    )?;
+
+    // Attempt to load and collect remaining non-fee payer accounts.
+    for (account_index, account_key) in account_keys.iter().enumerate().skip(1) {
+        let loaded_account = load_transaction_account(
+            account_loader,
+            message,
+            account_key,
+            account_index,
+            rent_collector,
+        );
+        collect_loaded_account(account_loader, account_key, loaded_account)?;
+    }
+
+    for (program_id, instruction) in message.program_instructions_iter() {
+        if native_loader::check_id(program_id) {
+            loaded_transaction_accounts.program_indices.push(vec![]);
+            continue;
+        }
+
+        let Some(program_account) = account_loader.load_account(program_id) else {
+            error_metrics.account_not_found += 1;
+            return Err(TransactionError::ProgramAccountNotFound);
+        };
+
+        if !account_loader
+            .feature_set
+            .remove_accounts_executable_flag_checks
+            && !program_account.executable()
+        {
+            error_metrics.invalid_program_for_execution += 1;
+            return Err(TransactionError::InvalidProgramForExecution);
+        }
+
+        let owner_id = program_account.owner();
+        if !native_loader::check_id(owner_id) && !PROGRAM_OWNERS.contains(owner_id) {
+            error_metrics.invalid_program_for_execution += 1;
+            return Err(TransactionError::InvalidProgramForExecution);
+        }
+
+        loaded_transaction_accounts
+            .program_indices
+            .push(vec![instruction.program_id_index as IndexOfAccount]);
```

Comment:
If we remove the `native_loader::check_id(program_id)` special case above, then we also don't need a `Vec` around the program index here anymore.

Reply:
we can do this as part of the feature gate removal, for now the old code still needs the vec

Reply:
Yes, exactly.